src/split_reduce.cpp

```diff
@@ -0,0 +1,208 @@
+    }
+
+    std::string name() const { return "split_fused_reduce"; }
+};
+MIGRAPHX_REGISTER_OP(split_fused_reduce);
+
+static bool is_reduce(const instruction& ins) { return contains(ins.name(), "reduce"); }
+
+static std::vector<instruction_ref> find_split(const_module_ref rm)
+{
+    std::vector<instruction_ref> result;
+    auto reduce_ins = std::find_if(rm->begin(), rm->end(), &is_reduce);
+    if(reduce_ins == rm->end())
+        return result;
+    // Bail if there is more than one reduce for now
+    if(std::any_of(std::next(reduce_ins), rm->end(), &is_reduce))
+        return result;
+    // Only handle reduce_sum for now
+    if(reduce_ins->name() != "reduce_sum")
+        return result;
+    result.push_back(reduce_ins);
+    // TODO: Find instructions that are used again in the module
+    return result;
+}
+
+static std::vector<instruction_ref> get_alive(const_module_ref rm,
+                                              const std::vector<instruction_ref>& splits)
+{
+    std::vector<instruction_ref> result;
+    bool stop = false;
+    liveness(*rm, [&](auto ins, const auto& live_set) {
+        if(stop)
+            return;
+        if(not contains(splits, ins))
+            return;
+        std::copy_if(live_set.begin(),
+                     live_set.end(),
+                     std::back_inserter(result),
+                     [](instruction_ref live) { return live->name() != "@param"; });
+        stop = true;
+    });
+    return result;
+}
+
+static std::string assign_op(const std::vector<instruction_ref>& splits)
+{
+    static std::unordered_map<std::string, std::string> m = {
+        {"reduce_sum", "assign_add"},
+        {"reduce_mean", "assign_add"},
+        {"reduce_prod", "assign_mul"},
+        {"reduce_max", "assign_max"},
+        {"reduce_min", "assign_min"},
+    };
+    return m.at(splits.front()->name());
+}
+
+static std::vector<instruction_ref>
+insert_module_inline(module& m, instruction_ref ins, const module::with_inputs& mwi)
+{
+    auto param_map = mwi.mod.get_ins_param_map(mwi.inputs, true);
+    return m.insert_instructions(ins, &mwi.mod, &param_map);
+}
+
+static std::size_t get_reduce_size(const_module_ref rm)
+{
+    auto ins = std::find_if(rm->begin(), rm->end(), &is_reduce);
+    assert(ins != rm->end());
+    return ins->inputs().front()->get_shape().elements() / ins->get_shape().elements();
+}
+
+void split_reduce::apply(module_pass_manager& mpm) const
+{
+    for(auto ins : iterator_for(mpm.get_module()))
+    {
+        if(ins->name() != "fused_reduce")
+            continue;
+        auto* rm = ins->module_inputs().front();
+        if(get_reduce_size(rm) < split_size)
+            continue;
+        auto splits = find_split(rm);
+        if(splits.empty())
+            continue;
+        // Only use split reduce with float for now
+        if(not std::all_of(splits.begin(), splits.end(), [](instruction_ref split) {
```

Comment:
Why only use for float? If to reduce scope for now, add a TODO here.

Reply:
The atomic instruction works directly with float. It doesnt work for half, it only works for half2, but to support that will need some non-trivial work.

Reply:
understood.